ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/KeyConstraints.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.planner.plan;
+
+import io.confluent.ksql.GenericKey;
+import io.confluent.ksql.planner.plan.PullFilterNode.WindowBounds;
+import java.util.Optional;
+
+public class KeyConstraints {
+
+  /**
+   * An interface for a KeyConstraint, which represents what we know about the use of keys in
+   * a given expression.  If there's a key value, an operator associated with it, or window bounds,
+   * they're available through the given methods. These are used as hints for the physical planning
+   * layer about how to fetch the corresponding rows.
+   */
+  public interface KeyConstraint {
```

Comment:
Yes I'm thinking about non-key lookups in the future. But I was not thinking about calling it `SecondaryKey`, since secondary-key indicates that there's a secondary-index available. But that might not be the case in practice. E.g. a constraint `C > 10` where `C` is not a key, and there's no secondary-index on `C`, would need to be translated as a scan still, calling it `SecondaryKeyLookup` sounds a bit weird.

> I was just trying to avoid cluttering the package with really simple classes, but I can just as well create a new subpackage if that's preferable.

Thanks! Just trying to understanding if we have any coding style guidance for such cases. For this one specifically I'm okay to keep it as-is.